growcrops.lua

```diff
@@ -0,0 +1,137 @@
+-- e.g. containsRefType(item.general_refs, df.general_ref_building_holderst)
+function containsRefType(refs, type)
+    for i, ref in ipairs(refs) do
+        if ref._type == type then
+            return true
+        end
+    end
+
+    return false
+end
+
+-- create a table of planted seed counts indexed by mat_index
+function collectInventory(crop_raws)
+    local inventory = {}
+    for _, seed in ipairs(df.global.world.items.other.SEEDS) do
+        -- skip if seed is not in a specific building (plot)
+        if not containsRefType(seed.general_refs, df.general_ref_building_holderst) then
+            goto skipseed
+        end
+
+        if seed.grow_counter < crop_raws[seed.mat_index].growdur then
+            inventory[seed.mat_index] = (inventory[seed.mat_index] or 0) + 1
+        end
+
+        :: skipseed ::
+    end
+
+    return inventory
+end
+
+-- Display a list of planted crops
+function listPlantedCrops(crop_raws, inventory)
+    local crops = {}
+    -- create a copy of crops from inventory to use for sorting
+    for mat, invCount in pairs(inventory) do
+        table.insert(crops, { name = crop_raws[mat].name, count = invCount })
+    end
+
+    -- sort the table in descending order by seed count
+    table.sort(crops, function(a, b) return a.count > b.count end)
+
+    for _, crop in ipairs(crops) do
+        print(string.format("%s %d", crop.name, crop.count))
+    end
+end
+
+-- Tries to find a crop raw in crop_raws that matches the given name
+-- The returned value contains the mat_index and name of the crop
+function findRawsCrop(crop_raws, material)
+    if not material then
+        return nil
+    end
+
+    local possible_matches = {}
+    local lower_mat = string.lower(material)
+    for mat_index, crop in pairs(crop_raws) do
+        local lower_crop_name = string.lower(crop.name)
+        if lower_mat == lower_crop_name then
+            return {mat_index = mat_index, name = crop.name}
+        end
+
+        if string.find(lower_crop_name, lower_mat) then
+            table.insert(possible_matches, {mat_index = mat_index, name = crop.name})
+        end
+    end
+
+    if #possible_matches == 1 then
+        return possible_matches[1]
+    end
+
+    return nil
+end
+
+-- grow specific crop
+function growCrop(crop_raws, material)
+    -- find the matching crop
+    local wanted_mat = findRawsCrop(crop_raws, material)
+    if not wanted_mat then
+        -- no crop with that name
+        qerror("invalid plant material " .. material)
+    end
+
+    -- grow each seed for specified crop
+    local count = 0
+    for _, seed in pairs(df.global.world.items.other.SEEDS) do
+        -- Skip seeds that don't match the material
+        if seed.mat_index ~= wanted_mat.mat_index then
+            goto growskip
+        end
+
+        -- skip if seed is not in a specific building (plot)
+        if not containsRefType(seed.general_refs, df.general_ref_building_holderst) then
+            goto growskip
+        end
```

Comment:
this might need more conditions -- maybe ensuring that the containing building actually *is* a farm plot? seeds could be in built bags or cages (when left over from animal training), for example.

Reply:
I agree, I just don't know how to do that. I wasn't seeing any existing examples of doing something similar, and the original script doesn't appear to handle that either.

Will look into it more deeply

Reply:
advfort does some things with general refs that might give you ideas:
https://github.com/DFHack/scripts/blob/727e4921c00e260d7c8d1112daf77115ce3960ee/gui/advfort.lua#L280